### PR TITLE
Feature/condo/doma 3372/role for contact

### PR DESCRIPTION
**Pull request by AleX83Xpert**
- [x] creating of roles
- [x] assign role to contact

### CODE REVIEW

File: apps/condo/domains/contact/access/ContactRole.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+const { getById } = require('@core/keystone/schema')
+const { STAFF } = require('@condo/domains/user/constants/common')
+
+async function canReadContactRoles ({ authentication: { item: user } }) {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+
+    if (user.isAdmin || user.isSupport) return {}
+
+    if (user.type !== STAFF) return false
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/access/ContactRole.js:21]: don't forget that we have a service user type

File: apps/condo/domains/contact/access/ContactRole.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    return {
+        OR: [
+            { organization_is_null: true },
+            { organization: queryOrganizationEmployeeFor(user.id) },
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/access/ContactRole.js:26]: what about employees from related `from` organization?
they have access to read and manage the contacts of the organization

File: apps/condo/domains/contact/access/ContactRole.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        if (!itemId) return false
+        organizationId = get(originalInput, ['organization', 'connect', 'id'])
+
+        if (!organizationId) {
+            const contactRole = await getById('ContactRole', itemId)
+            organizationId = get(contactRole, 'organization')
+        }
+    }
+
+    return checkOrganizationPermission(user.id, organizationId, 'canManageContacts')
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/access/ContactRole.js:51]: what about employees from related from organization?

File: apps/condo/domains/contact/access/ContactRole.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        ],
+    }
+}
+
+async function canManageContactRoles ({ authentication: { item: user }, originalInput, operation, itemId }) {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+    if (user.isAdmin || user.isSupport) return true
+
+    if (user.type !== STAFF) return false
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/access/ContactRole.js:36]: are we not giving access to users with the `service` type?

File: apps/condo/domains/contact/components/CreateContactForm.tsx [hunk 5]
```diff
@@ -53,6 +55,7 @@ export const CreateContactForm: React.FC = () => {
     const UnitLabel = intl.formatMessage({ id: 'field.Unit' })
     const PropertyErrorMessage = intl.formatMessage({ id: 'field.Property.requiredError' })
     const UnitErrorMessage = intl.formatMessage({ id: 'field.Unit.requiredError' })
+    const roleLabel = intl.formatMessage({ id: 'ContactRole' })
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/components/CreateContactForm.tsx:58]:
```suggestion
    const RoleLabel = intl.formatMessage({ id: 'ContactRole' })
```

File: apps/condo/domains/contact/access/ContactRole.js (hunk 1 above)
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/access/ContactRole.js:21]: are we not giving access to users with the `service` type?

File: apps/condo/domains/contact/access/ContactRole.js (hunk 2 above)
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/access/ContactRole.js:26]: what about employees from `related` from organization?
they have access to read and manage the contacts of the organization

File: apps/condo/domains/contact/access/ContactRole.js (hunk 4 above)
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/access/ContactRole.js:36]: what about user with `service` type?

File: apps/condo/domains/contact/access/ContactRole.js (hunk 3 above)
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/access/ContactRole.js:51]: i think we need to use `checkPermissionInUserOrganizationOrRelatedOrganization` to give manage access to employees in related `from` organization, how it is done in `access/Contact`

File: apps/condo/domains/contact/components/CreateContactForm.tsx (hunk 5 above)
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/components/CreateContactForm.tsx:58]:
```suggestion
    const RoleLabel = intl.formatMessage({ id: 'ContactRole' })
```

File: apps/condo/domains/contact/components/CreateContactForm.tsx [hunk 6]
```diff
@@ -75,6 +78,19 @@ export const CreateContactForm: React.FC = () => {
...
 
+    const {
+        loading: isRolesLoading,
+        count: totalRoles,
+        objs: roles,
+    } = ContactRole.useObjects({
+        where: {
+            OR: [
+                { organization_is_null: true },
+                { organization: { id: get(organization, 'id') } },
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/components/CreateContactForm.tsx:89]: please check that if `organization` is `undefined`, then we will not receive redundant contacts

File: apps/condo/domains/contact/components/EditContactForm.tsx [hunk 7]
```diff
@@ -37,13 +38,14 @@ export const EditContactForm: React.FC = () => {
...
     const NameLabel = intl.formatMessage({ id: 'field.FullName.short' })
-    const FullNamePlaceholderMessage = intl.formatMessage({ id:'field.FullName' })
+    const FullNamePlaceholderMessage = intl.formatMessage({ id: 'field.FullName' })
     const PhoneLabel = intl.formatMessage({ id: 'Phone' })
     const ExamplePhoneMessage = intl.formatMessage({ id: 'example.Phone' })
     const ExampleEmailMessage = intl.formatMessage({ id: 'example.Email' })
     const EmailLabel = intl.formatMessage({ id: 'field.EMail' })
     const ApplyChangesMessage = intl.formatMessage({ id: 'ApplyChanges' })
     const NoPermissionMessage = intl.formatMessage({ id: 'EditingContactNoPermission' })
+    const roleLabel = intl.formatMessage({ id: 'ContactRole' })
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/components/EditContactForm.tsx:48]:
```suggestion
    const RoleLabel = intl.formatMessage({ id: 'ContactRole' })
```

File: apps/condo/domains/contact/components/contactRoles/ContactRoleCard.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+const TEXT_STYLES: CSSProperties = { color: colors.black }
+const LINK_STYLES: CSSProperties = { ...TEXT_STYLES, position: 'relative', bottom: '-7px', fontSize: fontSizes.content }
+
+type TicketPropertyHintCardProps = {
+    propertyId: string
+    hintContentStyle?: CSSProperties
+}
+
+export const ContactRoleCard: React.FC<TicketPropertyHintCardProps> = ({ propertyId, hintContentStyle }) => {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/components/contactRoles/ContactRoleCard.tsx:36]: This component is identical to `TicketPropertyHintCard` and is not used anywhere, I think we need to delete this file

File: apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  .ant-table-cell-ellipsis {
+    white-space: inherit;
+  }
+`
+
+const MEDIUM_VERTICAL_GUTTER: [Gutter, Gutter] = [0, 40]
+
+export const ContactRolesSettingsContent = (props) => {
+    const intl = useIntl()
+    const titleMessage = intl.formatMessage({ id: 'ContactRoles' })
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx:32]:
```suggestion
    const TitleMessage = intl.formatMessage({ id: 'ContactRoles' })
```

File: apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    const addMessage = intl.formatMessage({ id: 'ContactRoles.add' })
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx:33]:
```suggestion
    const AddMessage = intl.formatMessage({ id: 'ContactRoles.add' })
```

File: apps/condo/domains/contact/access/ContactRole.js (hunk 1 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/access/ContactRole.js:21]: fixed

File: apps/condo/domains/contact/access/ContactRole.js (hunk 2 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/access/ContactRole.js:26]: done

File: apps/condo/domains/contact/access/ContactRole.js (hunk 4 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/access/ContactRole.js:36]: fixed

File: apps/condo/domains/contact/access/ContactRole.js (hunk 3 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/access/ContactRole.js:51]: > i think we need to use `checkPermissionInUserOrganizationOrRelatedOrganization` to give manage access to employees in related `from` organization, how it is done in `access/Contact`

done

> Are we not going to make a special role like `canManageContactRoles`?

I think that editing roles without editing contacts is not make sense.

File: apps/condo/domains/contact/components/CreateContactForm.tsx (hunk 5 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/CreateContactForm.tsx:58]: Why? This is just local variable.

File: apps/condo/domains/contact/components/CreateContactForm.tsx (hunk 6 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/CreateContactForm.tsx:89]: This query receives contact roles. In the case organization is undefined we receive just common roles.

File: apps/condo/domains/contact/components/EditContactForm.tsx (hunk 7 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/EditContactForm.tsx:48]: Local variable. It's good to use camelCase.

File: apps/condo/domains/contact/components/contactRoles/ContactRoleCard.tsx (hunk 8 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/contactRoles/ContactRoleCard.tsx:36]: Deleted :)

File: apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx (hunk 9 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx:32]: Keep camelCase for local variable

File: apps/condo/pages/contact/[id]/index.tsx [hunk 11]
```diff
@@ -50,6 +50,7 @@ export const ContactPageContent = ({ contact, isContactEditable, softDeleteActio
     const ConfirmDeleteButtonLabel = intl.formatMessage({ id: 'Delete' })
     const ConfirmDeleteTitle = intl.formatMessage({ id: 'contact.ConfirmDeleteTitle' })
     const ConfirmDeleteMessage = intl.formatMessage({ id: 'contact.ConfirmDeleteMessage' })
+    const contactRoleTitle = intl.formatMessage({ id: 'ContactRole' })
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/pages/contact/[id]/index.tsx:53]: Refactored to camelCase

File: apps/condo/domains/contact/access/ContactRole.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        if (!itemId) return false
+        organizationId = get(originalInput, ['organization', 'connect', 'id'])
+
+        if (!organizationId) {
+            const contactRole = await getById('ContactRole', itemId)
+            organizationId = get(contactRole, 'organization')
+        }
+    }
+
+    return checkPermissionInUserOrganizationOrRelatedOrganization(user.id, organizationId, 'canManageContacts')
```
Comment (pahaz) [inline apps/condo/domains/contact/access/ContactRole.js:47]: probably we can create a new `canManageContactRoles` permission

Comment (pahaz) [reply apps/condo/domains/contact/access/ContactRole.js:47]: It's looks different access level to manage contacts and manage their roles

File: apps/condo/domains/contact/components/CreateContactForm.tsx (hunk 5 above)
Comment (pahaz) [reply apps/condo/domains/contact/components/CreateContactForm.tsx:58]: we use CamelCase for translation variables. Do you really want to change this style?

File: apps/condo/domains/contact/components/CreateContactForm.tsx [hunk 13]
```diff
@@ -169,38 +191,54 @@ export const CreateContactForm: React.FC = () => {
...
                                             rules={validations.name}>
-                                            <Input placeholder={FullNamePlaceholderMessage}/>
+                                            <Input placeholder={fullNamePlaceholderMessage}/>
                                         </Form.Item>
                                     </Col>
                                     <Col lg={18} xs={24}>
                                         <Form.Item
                                             name={'phone'}
-                                            label={PhoneLabel}
+                                            label={phoneLabel}
```
Comment (pahaz) [inline apps/condo/domains/contact/components/CreateContactForm.tsx:206]: are you sure that you want to change the style of translation variables?

File: apps/condo/domains/contact/components/EditContactForm.tsx (hunk 7 above)
Comment (pahaz) [reply apps/condo/domains/contact/components/EditContactForm.tsx:48]: We already use CamalCase for translations in many place. Why you want to change this style?

File: apps/condo/pages/settings/index.tsx [hunk 14]
```diff
@@ -1,51 +1,73 @@
...
+import React, { CSSProperties, useMemo } from 'react'
+import { useOrganization } from '@core/next/organization'
 
 const TITLE_STYLES: CSSProperties = { margin: 0 }
 
-const ALWAYS_AVAILABLE_TABS = ['hint']
+const ALWAYS_AVAILABLE_TABS = [SETTINGS_TAB_PROPERTY_HINT]
 
-const SettingsPage = () => {
+const SettingsPage = (): JSX.Element => {
```
Comment (SavelevMatthew) [inline apps/condo/pages/settings/index.tsx:28]: `React.FC`

File: apps/condo/domains/contact/components/CreateContactForm.tsx (hunk 5 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/CreateContactForm.tsx]: ok, commits reverted. New translation variables renamed in ~~Pascal~~CamelCase.

File: apps/condo/domains/contact/access/ContactRole.js (hunk 3 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/access/ContactRole.js]: `canManageContactRoles` implemented

File: apps/condo/domains/contact/access/ContactRole.js (hunk 12 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/access/ContactRole.js]: done

File: apps/condo/domains/contact/components/contactRoles/BaseContactRoleForm.tsx [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    hintFilters,
+    children,
+    action,
+    organizationId,
+    initialValues,
+    mode,
+}) => {
+    const intl = useIntl()
+    const NameMessage = intl.formatMessage({ id: 'ContactRoles.name' })
+    const router = useRouter()
```
Comment (pahaz) [inline apps/condo/domains/contact/components/contactRoles/BaseContactRoleForm.tsx:45]:
```suggestion

    const router = useRouter()
```

File: apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx (hunk 10 above)
Comment (pahaz) [inline apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx:33]:
```suggestion
    const AddMessage = intl.formatMessage({ id: 'ContactRoles.add' })
```

File: apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+            {
+                canManageContacts && (
+                    <Col span={24}>
+                        <ActionBar>
+                            <Button
+                                type={'sberDefaultGradient'}
+                                icon={<PlusCircleOutlined/>}
+                                onClick={handleAddHintButtonClick}
+                            >
+                                {addMessage}
```
Comment (pahaz) [inline apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx:108]:
```suggestion
                                {AddMessage}
```

File: apps/condo/domains/contact/schema/ContactRole.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+const ContactRole = new GQLListSchema('ContactRole', {
+    schemaDoc: 'Role for contact',
+    fields: {
+        organization: {
+            schemaDoc: 'The organization that created this role. Null means role for all organizations.',
+            type: Relationship,
+            ref: 'Organization',
+            isRequired: false,
+            knexOptions: { isNotNullable: false }, // Relationship only!
+            kmigratorOptions: { null: true, on_delete: 'models.CASCADE' },
```
[the remark this answers is not in the document]
Comment (pahaz) [reply apps/condo/domains/contact/schema/ContactRole.js:21]: CASCADE is a good for this case

**Review by pahaz — APPROVED** [review]

File: apps/condo/domains/contact/access/ContactRole.js [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+async function canReadContactRoles ({ authentication: { item: user } }) {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+
+    if (user.isAdmin || user.isSupport) return {}
+
+    return {
+        OR: [
+            { organization_is_null: true },
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/contact/access/ContactRole.js:22]: This part of the query loads roles with `organization=null` (common roles)

File: apps/condo/domains/contact/components/EditContactForm.tsx [hunk 19]
```diff
@@ -63,10 +65,22 @@ export const EditContactForm: React.FC = () => {
...
-        })
+    const {
+        loading: isRolesLoading,
+        count: totalRoles,
+        objs: roles,
+    } = ContactRole.useObjects({
+        where: {
+            OR: [
+                { organization_is_null: true },
+                { organization: { id: get(organization, 'id') } },
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/EditContactForm.tsx:76]: Fixed

File: apps/condo/pages/settings/contactRole/[id]/update.tsx [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+                <title>{PageTitleMsg}</title>
+            </Head>
+            <PageWrapper>
+                <PageContent>
+                    <Row gutter={BIG_VERTICAL_GUTTER} style={ROW_STYLES}>
+                        <Col span={24}>
+                            <Typography.Title level={1} style={TITLE_STYLES}>{PageTitleMsg}</Typography.Title>
+                        </Col>
+                        <Col span={24}>
+                            <ContactRoleForm id={query.id as string}/>
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/pages/settings/contactRole/[id]/update.tsx:32]: `id` sent by Nextjs. This id got from path.
image

File: apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx (hunk 10 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx:33]: fixed

File: apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx (hunk 16 above)
Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/contactRoles/ContactRolesSettingsContent.tsx:108]: fixed

File: apps/condo/domains/contact/components/EditContactForm.tsx [hunk 21]
```diff
@@ -99,6 +113,7 @@ export const EditContactForm: React.FC = () => {
         name: get(contact, 'name'),
         phone: get(contact, 'phone'),
         email: get(contact, 'email'),
+        role: get(contact, 'role.id'),
```
Comment (SavelevMatthew) [inline apps/condo/domains/contact/components/EditContactForm.tsx:116]:
```suggestion
        role: get(contact, ['role', 'id']),
```

Comment (AleX83Xpert) [reply apps/condo/domains/contact/components/EditContactForm.tsx]: Absolutely the same. But fixed.

File: apps/condo/migrations/20220721114406-0151_contactrolehistoryrecord_contacthistoryrecord_role_and_more.js [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+CREATE TABLE "ContactRoleHistoryRecord" ("organization" uuid NULL, "name" text NULL, "id" uuid NOT NULL PRIMARY KEY, "v" integer NULL, "createdAt" timestamp with time zone NULL, "updatedAt" timestamp with time zone NULL, "createdBy" uuid NULL, "updatedBy" uuid NULL, "deletedAt" timestamp with time zone NULL, "newId" jsonb NULL, "dv" integer NULL, "sender" jsonb NULL, "history_date" timestamp with time zone NOT NULL, "history_action" varchar(50) NOT NULL, "history_id" uuid NOT NULL);
+--
+-- Add field role to contacthistoryrecord
+--
+ALTER TABLE "ContactHistoryRecord" ADD COLUMN "role" uuid NULL;
+--
+-- Add field canManageContactRoles to organizationemployeerole
+--
+ALTER TABLE "OrganizationEmployeeRole" ADD COLUMN "canManageContactRoles" boolean DEFAULT false NOT NULL;
+ALTER TABLE "OrganizationEmployeeRole" ALTER COLUMN "canManageContactRoles" DROP DEFAULT;
```
Comment (nomerdvadcatpyat) [inline apps/condo/migrations/20220721114406-0151_contactrolehistoryrecord_contacthistoryrecord_role_and_more.js:19]: You need to specify which of the default roles can manage contact roles, add something like this:
`UPDATE "OrganizationEmployeeRole" SET "canManageContactRoles" = true WHERE "name" = ANY(ARRAY['employee.role.Dispatcher.name', 'employee.role.Administrator.name']);`

**Comment by nomerdvadcatpyat** [comment]
there is too much space between the title and the role "name" field
изображение

File: apps/condo/migrations/20220721114406-0151_contactrolehistoryrecord_contacthistoryrecord_role_and_more.js (hunk 22 above)
Comment (AleX83Xpert) [reply apps/condo/migrations/20220721114406-0151_contactrolehistoryrecord_contacthistoryrecord_role_and_more.js:19]: Done

**Comment by AleX83Xpert** [comment]
> there is too much space between the title and the role "name" field изображение

Fixed

**Review by nomerdvadcatpyat — APPROVED** [review]

**Review by SavelevMatthew — APPROVED** [review]

**Pull request merged**